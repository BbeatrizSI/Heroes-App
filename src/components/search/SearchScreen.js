import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import HeroCard from '../heroes/HeroCard';
import getHeroByName from '../../selectors/getHeroByName';


const SearchScreen = ({history}) => {
    const location = useLocation();
    const {q = ''} = queryString.parse(location.search);

    const [ formValues, handleInputChange ] = useForm ({searchText: q});
    const {searchText} = formValues;

    const heroesFiltered = useMemo(() => getHeroByName(q),[q])
        
    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${searchText}`);
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <hr/>
            <div className="row">
                <div className="col-5">
                    <h4>Search form</h4>
                    <hr/>
                    <form onSubmit={handleSearch}>
                        <input type="text" placeholder="Find your hero" className="form-control" name="searchText" value={ searchText } onChange={ handleInputChange } autoComplete="off" required/>
                        <button type="submit" className="btn m-1 btn-block btn-outline-primary">Search</button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr/>

                    {
                        (q === '') && <div className="alert alert-info">
                            Search a hero
                        </div>
                    }
                    {
                        (q !== '' && heroesFiltered.length === 0) && <div className="alert alert-danger">
                            There is no hero with {q}
                        </div>
                    }
                    {
                        heroesFiltered.map( hero => (
                            <HeroCard key={hero.id} {...hero} />
                        ))
                    }

                </div>  
            </div>
        </div>
    );
};

export default SearchScreen;