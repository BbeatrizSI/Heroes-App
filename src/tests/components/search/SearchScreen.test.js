import React from 'react';
import { mount } from 'enzyme';
import  SearchScreen  from '../../../components/search/SearchScreen';
import { MemoryRouter, Route } from 'react-router-dom';



describe('Test in <SearchScreen />', () => {

    test('should show correctly with default values', () => {
        
       const wrapper = mount(
           <MemoryRouter initialEntries={['/search']}>
               <Route path="/search" component={ SearchScreen }/>
           </MemoryRouter>
       )

       expect( wrapper ).toMatchSnapshot();
       expect( wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
                     
    });

    test('should show Batman and input with queryString value', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={ SearchScreen }/>
            </MemoryRouter>
        )
 
       expect( wrapper.find('input').prop('value')).toBe('batman');
       expect( wrapper ).toMatchSnapshot();
                      
     });

    test('should show error if hero isnt found', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Route path="/search" component={ SearchScreen }/>
            </MemoryRouter>
        )
 
        expect( wrapper.find('.alert-danger').text().trim()).toBe('There is no hero with batman123');
                      
    });
    test('should call history push', () => {
        
        const history = {
            push: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Route path="/search" component={ () => <SearchScreen history={ history } /> } />
            </MemoryRouter>
        )
 
        wrapper.find('input').simulate('change', {
           target: {
               name: 'searchText',
               value: 'batman'
           } 
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });
        
        expect( history.push ).toHaveBeenCalledWith(`?q=batman`)
        
    });

    
})