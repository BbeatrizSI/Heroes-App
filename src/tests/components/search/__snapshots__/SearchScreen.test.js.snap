// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <SearchScreen /> should show Batman and input with queryString value 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test in <SearchScreen /> should show correctly with default values 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
